src/mikro-orm-core.module.ts

```diff
@@ -49,73 +57,80 @@ export class MikroOrmCoreModule implements OnApplicationShutdown {
+        ...(await whenModuleAvailable(EntityManagerModuleName.MongoDb, ({ MongoEntityManager })  => contextName ? getMongoEntityManagerToken(contextName) : MongoEntityManager)),
       ],
     };
   }
 
   static async forRootAsync(options: MikroOrmModuleAsyncOptions): Promise<DynamicModule> {
+    const contextName = this.setContextName(options?.contextName);
     return {
       module: MikroOrmCoreModule,
       imports: options.imports || [],
       providers: [
         ...(options.providers || []),
-        ...createAsyncProviders(options),
-        createMikroOrmProvider(),
-        createMikroOrmEntityManagerProvider(options.scope),
-        ...(await whenModuleAvailable(EntityManagerModuleName.Knex, ({ SqlEntityManager })  => createMikroOrmEntityManagerProvider(options.scope, SqlEntityManager))),
-        ...(await whenModuleAvailable(EntityManagerModuleName.MongoDb, ({ MongoEntityManager })  => createMikroOrmEntityManagerProvider(options.scope, MongoEntityManager))),
+        ...createAsyncProviders({ ...options, contextName: options.contextName }),
+        createMikroOrmProvider(contextName),
+        createMikroOrmEntityManagerProvider(options.scope, EntityManager, contextName),
+        ...(await whenModuleAvailable(EntityManagerModuleName.Knex, ({ SqlEntityManager })  => createMikroOrmEntityManagerProvider(options?.scope, contextName ? getSqlEntityManagerToken(contextName) : SqlEntityManager, contextName))),
+        ...(await whenModuleAvailable(EntityManagerModuleName.MongoDb, ({ MongoEntityManager })  => createMikroOrmEntityManagerProvider(options?.scope, contextName ? getMongoEntityManagerToken(contextName) : MongoEntityManager, contextName))),
       ],
       exports: [
-        MikroORM,
-        EntityManager,
-        ...(await whenModuleAvailable(EntityManagerModuleName.Knex, ({ SqlEntityManager })  => SqlEntityManager)),
-        ...(await whenModuleAvailable(EntityManagerModuleName.MongoDb, ({ MongoEntityManager })  => MongoEntityManager)),
+        contextName ? getMikroORMToken(contextName) : MikroORM,
+        contextName ? getEntityManagerToken(contextName) : EntityManager,
+        ...(await whenModuleAvailable(EntityManagerModuleName.Knex, ({ SqlEntityManager })  => contextName ? getSqlEntityManagerToken(contextName) : SqlEntityManager)),
+        ...(await whenModuleAvailable(EntityManagerModuleName.MongoDb, ({ MongoEntityManager })  => contextName ? getMongoEntityManagerToken(contextName) : MongoEntityManager)),
       ],
     };
   }
 
   async onApplicationShutdown() {
-    const orm = this.moduleRef.get(MikroORM);
+    const token = this.options.contextName ? getMikroORMToken(this.options.contextName) : MikroORM;
+    const orm = this.moduleRef.get(token);
 
     if (orm) {
       await orm.close();
     }
+
+    CONTEXT_NAMES.splice(0,CONTEXT_NAMES.length);
```

Comment:
this is supposed to be clearning the array, right? if so:

```suggestion
    CONTEXT_NAMES.length = 0;
```

maybe we could use a `Set` instead of array